openlibrary/plugins/upstream/table_of_contents.py

```diff
@@ -120,23 +124,30 @@ def from_markdown(line: str) -> 'TocEntry':
         >>> f("1.1 | Apple")
         (0, '1.1', 'Apple', None)
         """
+
         RE_LEVEL = web.re_compile(r"(\**)(.*)")
         level, text = RE_LEVEL.match(line.strip()).groups()
 
         if "|" in text:
             tokens = text.split("|", 3)
             label, title, page, extra_fields = pad(tokens, 4, '')
+            if "|" in tokens[-1]:
+                raise TocParseError("Too many pipes!")
         else:
             title = text
             label = page = ""
             extra_fields = ''
+        extra_fields = json.loads(extra_fields or '{}')
```

Comment:
We can also handle JSON errors here, which would be nice!
```suggestion
        try:
            extra_fields = json.loads(extra_fields or '{}')
        except json.decoder.JSONDecodeError as e:
            raise TocParseError(f"Invalid JSON: {e.message}")
```